bridge/sam/pool.go

```diff
@@ -0,0 +1,399 @@
+		p.tryToPromoteAndDemoteAll()
+	} else if len(maybeDemotableHashes) > 0 {
+		for _, hash := range maybeDemotableHashes {
+			p.tryToDemote(hash)
+		}
+	}
+}
+
+// canPromote return the flag indicating it's possible to change status to ready
+// message need to have enough signatures and be known by pool for promotion
+func (p *pool) canPromote(hash types.Hash) bool {
+	isKnown := p.knows(hash)
+	numSignatures := p.messageSignatures.GetSignatureCount(hash)
+	threshold := atomic.LoadUint64(&p.threshold)
+
+	return isKnown && numSignatures >= threshold
+}
+
+// canDemote return the flag indicating it's possible to change status to pending
+func (p *pool) canDemote(hash types.Hash) bool {
+	return !p.canPromote(hash)
+}
+
+// tryToPromote checks the number of signatures and threshold and update message status to ready if need
+func (p *pool) tryToPromote(hash types.Hash) {
+	fmt.Printf("tryToPromote %v\n", hash)
+	if p.canPromote(hash) {
+		p.promote(hash)
+	}
+}
+
+// tryToDemote checks the number of signatures and threshold and update message status to pending if need
+func (p *pool) tryToDemote(hash types.Hash) {
+	if p.canDemote(hash) {
+		p.demote(hash)
+	}
+}
+
+// tryToPromoteAndDemoteAll iterates all messages and update its statuses
+func (p *pool) tryToPromoteAndDemoteAll() {
+	threshold := atomic.LoadUint64(&p.threshold)
+
+	p.messageSignatures.RangeMessages(func(entry *signedMessageEntry) bool {
+		hash := entry.Hash
+		isKnown := p.knows(hash)
+		numSignatures := entry.NumSignatures()
+
+		if numSignatures >= threshold && isKnown {
+			p.promote(hash)
+		} else {
+			p.demote(hash)
+		}
+
+		return true
+	})
+}
+
+// promote change message status to ready
+func (p *pool) promote(hash types.Hash) {
+	p.readyMap.Store(hash, true)
+}
+
+// promote change message status to pending
+// it deletes instead of unsetting for less-complexity on getting ready messages
+func (p *pool) demote(hash types.Hash) {
+	p.readyMap.Delete(hash)
+}
+
+func (p *pool) setMessageBody(hash types.Hash, body []byte) {
+	p.messageMap.Store(hash, body)
+}
+
+func (p *pool) getMessageBody(hash types.Hash) []byte {
+	raw, existed := p.messageMap.Load(hash)
+	if !existed {
+		return nil
+	}
+
+	body, ok := raw.([]byte)
+	if !ok {
+		// should not reach
+
+		return nil
+	}
+
+	return body
+}
+
+// signedMessageEntry is representing the data stored in messageSignaturesStore
+type signedMessageEntry struct {
+	Hash           types.Hash
+	Signatures     sync.Map
+	SignatureCount int64
+}
+
+// NumSignatures returns number of signatures
+func (e *signedMessageEntry) NumSignatures() uint64 {
+	count := atomic.LoadInt64(&e.SignatureCount)
+	if count < 0 {
+		return 0
+	}
+
+	return uint64(count)
+}
+
+// IncrementNumSignatures increments SignatureCount and return new count
+func (e *signedMessageEntry) IncrementNumSignatures() uint64 {
+	newNumSignatures := atomic.AddInt64(&e.SignatureCount, 1)
+	if newNumSignatures < 0 {
+		return 0
+	}
```

Comment:
Is this even possible?

Reply:
Fixed at c92a67bf10bc73bcdf8929dca389a8e53020bf04